File: tests/e2e/BaseTestCase.php
<?php

namespace tests\e2e;

/**
 * Class BaseTestCase
 *
 * Base TestCase that all e2e tests should extend as it provides a common method for logging into and out of the site,
 * as well as some useful utility functions for running the tests. Any class that extends this class should at a
 * minimum override the class variables $user_id, $password, and $test_url for that particular test case. If a custom
 * setUp or tearDown routine is defined within the child class, it should also still call the parent class's function
 * so as to ensure that logging in and out still happens (unless that's not necessary for a particular test)
 */
class BaseTestCase extends \PHPUnit_Extensions_Selenium2TestCase {
    /* These variables should be overwritten */
    /** @var string user id to use for logging into the site */
    protected $user_id = null;
    /** @var string password to use for logging into the site */
    protected $password = null;
    /** @var string URL to use as the base for the tests */
    protected $test_url = __TEST_URL__;
    
    private $logged_in = false;
    
    // For every browser, always specify sessionStrategy to be shared
    // as this causes one browser window per test class as opposed to
    // one browser window per test function (which is way slower)
    public static $browsers = array(
        array(
            'browserName' => __BROWSER__,
            'sessionStrategy' => 'shared'
        )
    );

    public function logIn($user_id, $password) {
        $this->url("/index.php?semester=f16&course=csci1000");
        try {
            $this->byId("login-guest");
            $this->byName("user_id")->value($user_id);
            $this->byName("password")->value($password);
            $this->byName("login")->click();
        }
        catch (\PHPUnit_Extensions_Selenium2TestCase_WebDriverException $e) {
            // We're logged in already to the system, just ensure that we're logged in as the right user below
        }

        $time = $this->timeouts()->getLastImplicitWaitValue();
        $this->timeouts()->implicitWait(2500);
        try {
            $this->byId('login');
            $this->assertEquals($this->user_id, $this->byId('login-id')->text());
        }
        catch (\PHPUnit_Extensions_Selenium2TestCase_WebDriverException $e) {
            print(exec('whoami'));
            print($this->source());
            print($this->url());
        }
        $this->timeouts()->implicitWait($time);
        $this->logged_in = true;
    }

    /**
     * Log out of the site, making sure there's no open sessions (so to not affect the next test that gets
     * run which might need to be logged in as a different user). However, we only want to run this if we've
     * successfully logged in, else a WebDriverException will end up being thrown as the element won't be found.
     */
    public function logOut() {
        if ($this->logged_in) {
            $this->byId("logout")->click();
        }
    }

    /**
     * Function run before every test function, but we don't have access to session/selenium yet
     */
    public function setUp() {
        $this->setBrowserUrl($this->test_url);
    }

    /**
     * Generic set up method for the e2e tests. This function is run once after the session has been created
     * (so after setUp()) so that we can access the url() function.
     */
    public function setUpPage() {
        $this->logIn($this->user_id, $this->password);
    }

    /**
     * Function run after every test function, and we do have access to session/selenium details still
     */
    public function tearDown() {
        $this->logOut();
    }
    
    /**
     * This causes the selenium test to wait until the user hits enter on the PHP console that's running the test.
     * This shouldn't be used within any any actual tests, but can be useful for creating and debugging tests as it'll
     * cause the selenium window to pause indefinitely allowing for inspection of the HTML elements
     */
    protected function waitForUserInput() {
        if(trim(fgets(fopen("php://stdin","r"))) != chr(13)) {
            return;
        }
    }
}

File: tests/e2e/base_testcase.py
import shutil
import tempfile
from datetime import date
import os
import unittest

from urllib.parse import urlencode

from selenium import webdriver

from selenium.common.exceptions import WebDriverException
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By


# noinspection PyPep8Naming
class BaseTestCase(unittest.TestCase):
    """
    Base class that all e2e tests should extend. It provides several useful
    helper functions, sets up the selenium webdriver, and provides a common
    interface for logging in/out a user. Each test then only really needs to
    override user_id, user_name, and user_password as necessary for a
    particular testcase and this class will handle the rest to setup the test.
    """
    TEST_URL = "http://localhost:1501"
    USER_ID = "student"
    USER_NAME = "Joe"
    USER_PASSWORD = "student"

    WAIT_TIME = 20

    def __init__(self, testname, user_id=None, user_password=None, user_name=None, log_in=True):
        super().__init__(testname)
        if "TEST_URL" in os.environ and os.environ['TEST_URL'] is not None:
            self.test_url = os.environ['TEST_URL']
        else:
            self.test_url = BaseTestCase.TEST_URL
        self.driver = None
        """ :type driver: webdriver.Chrome """
        self.options = webdriver.ChromeOptions()
        self.options.add_argument('--no-sandbox')
        self.options.add_argument('--headless')
        self.options.add_argument("--disable-extensions")
        self.options.add_argument('--hide-scrollbars')
        self.options.add_argument('--disable-gpu')
        self.options.add_argument('--no-proxy-server')

        self.download_dir = tempfile.mkdtemp(prefix="vagrant-submitty")
        # https://stackoverflow.com/a/26916386/214063
        profile = {
            'download.prompt_for_download': False,
            'download.default_directory': self.download_dir,
            'download.directory_upgrade': True,
            'plugins.plugins_disabled': ['Chrome PDF Viewer']
        }
        self.options.add_experimental_option('prefs', profile)
        self.user_id = user_id if user_id is not None else BaseTestCase.USER_ID
        self.user_name = user_name if user_name is not None else BaseTestCase.USER_NAME
        if user_password is None and user_id is not None:
            user_password = user_id
        self.user_password = user_password if user_password is not None else BaseTestCase.USER_PASSWORD
        self.semester = BaseTestCase.get_current_semester()
        self.full_semester = BaseTestCase.get_display_semester(self.semester)
        self.logged_in = False
        self.use_log_in = log_in

    def setUp(self):
        # attempt to set-up the connection to Chrome. Repeat a handful of times
        # in-case Chrome crashes during initialization
        for _ in range(5):
            try:
                self.driver = webdriver.Chrome(options=self.options)
                break
            except WebDriverException:
                pass
        if self.driver is None:
            self.driver = webdriver.Chrome(options=self.options)

        self.driver.set_window_size(1600, 900)
        self.enable_download_in_headless_chrome(self.download_dir)
        if self.use_log_in:
            self.log_in()

    def tearDown(self):
        self.driver.quit()
        shutil.rmtree(self.download_dir)

    def get(self, url=None, parts=None):
        if url is None:
            # Can specify parts = [('semester', 's18'), ...]
            self.assertIsNotNone(parts)
            url = "/index.php?" + urlencode(parts)

        if url[0] != "/":
            url = "/" + url
        self.driver.get(self.test_url + url)

        # Frog robot
        self.assertNotEqual(self.driver.title, "Submitty - Error", "Got Error Page")

    def log_in(self, url=None, title="Submitty", user_id=None, user_password=None, user_name=None):
        """
        Provides a common function for logging into the site (and ensuring
        that we're logged in)
        :return:
        """
        if url is None:
            url = "/index.php"

        if user_password is None:
            user_password = user_id if user_id is not None else self.user_password
        if user_id is None:
            user_id = self.user_id
        if user_name is None:
            user_name = self.user_name

        self.get(url)
        # print(self.driver.page_source)
        self.assertIn(title, self.driver.title)
        self.driver.find_element(By.NAME, 'user_id').send_keys(user_id)
        self.driver.find_element(By.NAME, 'password').send_keys(user_password)
        self.driver.find_element(By.NAME, 'login').click()

        # OLD self.assertEqual(user_name, self.driver.find_element(By.ID, "login-id").get_attribute('innerText').strip(' \t\r\n'))

        # FIXME: WANT SOMETHING LIKE THIS...  WHEN WE HAVE JUST ONE ELEMENT WITH THIS ID
        # self.assertEqual("Logout "+user_name, self.driver.find_element(By.ID, "logout").get_attribute('innerText').strip(' \t\r\n'))

        # instead, just make sure this element exists
        self.driver.find_element(By.ID, "logout")

        self.logged_in = True

    def log_out(self):
        if self.logged_in:
            self.logged_in = False
            self.driver.find_element(By.ID, 'logout').click()
            self.driver.find_element(By.ID, 'login-guest')

    def click_class(self, course, course_name=None):
        if course_name is None:
            course_name = course
        course_name = course_name.title()
        self.driver.find_element(By.ID, self.get_current_semester() + '_' + course).click()
        # print(self.driver.page_source)
        WebDriverWait(self.driver, BaseTestCase.WAIT_TIME).until(EC.title_is('Submitty ' + course_name + ' Gradeables'))

    # see Navigation.twig for html attributes to use as arguments
    # loaded_selector must recognize an element on the page being loaded (test_simple_grader.py has xpath example)
    def click_nav_grade_button(self, gradeable_category, gradeable_id, button_name, loaded_selector):
        self.driver.find_element(By.XPATH,
            "//div[@id='{}']/div[@class='course-button']/a[contains(@class, 'btn-nav-grade')]".format(
                gradeable_id)).click()
        WebDriverWait(self.driver, BaseTestCase.WAIT_TIME).until(EC.presence_of_element_located(loaded_selector))

    def click_nav_submit_button(self, gradeable_category, gradeable_id, button_name, loaded_selector):
        self.driver.find_element(By.XPATH,
            "//div[@id='{}']/div[@class='course-button']/a[contains(@class, 'btn-nav-submit')]".format(
                gradeable_id)).click()
        WebDriverWait(self.driver, BaseTestCase.WAIT_TIME).until(EC.presence_of_element_located(loaded_selector))

    # clicks the navigation header text to 'go back' pages
    # for homepage, selector can be gradeable list
    def click_header_link_text(self, text, loaded_selector):
        self.driver.find_element(
            By.XPATH,
            "//div[@id='breadcrumbs']/div[@class='breadcrumb']/a[text()='{}']".format(text)
        ).click()
        WebDriverWait(self.driver, BaseTestCase.WAIT_TIME).until(EC.presence_of_element_located(loaded_selector))

    def wait_after_ajax(self):
        WebDriverWait(self.driver, 10).until(lambda driver: driver.execute_script("return jQuery.active == 0"))

    def wait_for_element(self, element_selector, visibility=True, timeout=WAIT_TIME):
        """
        Waits for an element to be present in the DOM. By default, also waits for the element to be
        visible/interactable
        """
        if visibility:
            WebDriverWait(self.driver, timeout).until(EC.visibility_of_element_located(element_selector))
        else:
            WebDriverWait(self.driver, timeout).until(EC.presence_of_element_located(element_selector))


    @staticmethod
    def wait_user_input():
        """
        Causes the running selenium test to pause until the user has hit the enter key in the
        terminal that is running python. This is useful for using in the middle of building tests
        as then you cna use the javascript console to inspect the page, get the name/id of elements
        or other such actions and then use that to continue building the test
        """
        input("Hit enter to continue...")

    @staticmethod
    def get_current_semester():
        """
        Returns the "current" academic semester which is in use on the Vagrant/Travis machine (as we
        want to keep referring to a url that is "up-to-date"). The semester will either be spring
        (prefix "s") if we're in the first half of the year otherwise fall (prefix "f") followed
        by the last two digits of the current year. Unless you know you're using a course that
        was specifically set-up for a certain semester, you should always be using the value
        generated by this function in the code.

        :return:
        """
        today = date.today()
        semester = "f" + str(today.year)[-2:]
        if today.month < 7:
            semester = "s" + str(today.year)[-2:]
        return semester

    @staticmethod
    def get_display_semester(current_semester):
        s = 'Fall' if current_semester[0] == 'f' else 'Summer' if current_semester[0] == 'u' else 'Spring'
        s += ' 20' + current_semester[1:]
        return s

    # https://stackoverflow.com/a/47366981/214063
    def enable_download_in_headless_chrome(self, download_dir):
        # add missing support for chrome "send_command"  to selenium webdriver
        self.driver.command_executor._commands["send_command"] = ("POST", '/session/$sessionId/chromium/send_command')

        params = {'cmd': 'Page.setDownloadBehavior', 'params': {'behavior': 'allow', 'downloadPath': download_dir}}
        self.driver.execute("send_command", params)
